### PR TITLE
Intake IO Layers

### PR DESCRIPTION
**Pull request by falOn-Dev**

### CODE REVIEW

File: src/main/java/frc/robot/subsystems/intake/IntakeIONeo.kt [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+    private val motorID: Int,
+    private val intakeSensorID: Int,
+    private val exitSensorID: Int
+) : IntakeIO {
+
+    private val motor = CANSparkMax(motorID, CANSparkLowLevel.MotorType.kBrushless).apply {
+        restoreFactoryDefaults()
+        idleMode = CANSparkBase.IdleMode.kBrake
+        inverted = false
+        setSmartCurrentLimit(20)
```
Comment (mteam88) [inline src/main/java/frc/robot/subsystems/intake/IntakeIONeo.kt]: What's this number from?

File: src/main/java/frc/robot/subsystems/intake/IntakeIONeo.kt [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+    }
+
+    private val exitSensor: DigitalInput = DigitalInput(exitSensorID)
+    private val intakeSensor: DigitalInput = DigitalInput(intakeSensorID)
+
+    override fun updateInputs(inputs: IntakeIO.IntakeInputs) {
+        inputs.intakeSensor = intakeSensor.get()
+        inputs.exitSensor = exitSensor.get()
+        inputs.intakeLinearVelocity = motor.encoder.velocity * ((Units.inchesToMeters(0.84) * PI)/ 60.0)
```
Comment (mteam88) [inline src/main/java/frc/robot/subsystems/intake/IntakeIONeo.kt]: What is this calculation?

File: src/main/java/frc/robot/subsystems/intake/IntakeIOSim.kt [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+import edu.wpi.first.wpilibj.RobotController
+import edu.wpi.first.wpilibj.simulation.FlywheelSim
+import lib.math.units.into
+
+class IntakeIOSim : IntakeIO {
+    private val intakeSim: FlywheelSim = FlywheelSim(
+        DCMotor.getNeo550(1),
+        2.0,
+        0.0000719
+    )
```
Comment (mteam88) [inline src/main/java/frc/robot/subsystems/intake/IntakeIOSim.kt]: What is all this?

File: src/main/java/frc/robot/subsystems/intake/IntakeIOSim.kt [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+
+    private val intakeSensorBool: SimBoolean = SimBoolean(1)
+    private val exitSensorBool: SimBoolean = SimBoolean(2)
+
+    private var cachedVoltage: MutableMeasure<Voltage> = MutableMeasure.zero(Volts)
+
+    override fun updateInputs(inputs: IntakeIO.IntakeInputs) {
+        inputs.intakeSensor = intakeSensorBool.get()
+        inputs.exitSensor = exitSensorBool.get()
+        inputs.intakeLinearVelocity = intakeSim.angularVelocityRadPerSec * (Units.inchesToMeters(0.84) / 2)
```
Comment (mteam88) [inline src/main/java/frc/robot/subsystems/intake/IntakeIOSim.kt]: What is this calculation?

File: src/main/java/frc/robot/subsystems/intake/IntakeIOSim.kt [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+        inputs.intakeSupplyVoltage = 12.0
```
Comment (mteam88) [inline src/main/java/frc/robot/subsystems/intake/IntakeIOSim.kt]: Where is this from?

File: src/main/java/frc/robot/subsystems/intake/IntakeIO.kt [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+import org.littletonrobotics.junction.inputs.LoggableInputs
+
+interface IntakeIO {
+    class IntakeInputs : LoggableInputs {
+        var intakeLinearVelocity: Double = 0.0
+        var intakeSupplyVoltage: Double = 0.0
+        var intakeMotorVoltage: Double = 0.0
+        var intakeStatorCurrent: Double = 0.0
+        var exitSensor: Boolean = false
+        var intakeSensor: Boolean = false
```
Comment (Something-Inconspicuous) [inline src/main/java/frc/robot/subsystems/intake/IntakeIO.kt]: These names are not that good. Is it just for one 20ms tick these are true? Their purpose should be made clear as booleans; currently they seem like they should be `DigitalInput`.

File: src/main/java/frc/robot/subsystems/intake/IntakeIO.kt [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+            intakeLinearVelocity = table.get("intakeLinearVelocity", intakeLinearVelocity)
+            intakeSupplyVoltage = table.get("intakeSupplyVoltage", intakeSupplyVoltage)
+            intakeMotorVoltage = table.get("intakeMotorVoltage", intakeMotorVoltage)
+            intakeStatorCurrent = table.get("intakeStatorCurrent", intakeStatorCurrent)
+            exitSensor = table.get("exitSensor", exitSensor)
+            intakeSensor = table.get("intakeSensor", intakeSensor)
+        }
+    }
+    fun updateInputs(inputs: IntakeInputs) {}
+    fun applyVoltage(voltage: Measure<Voltage>) {}
```
Comment (Something-Inconspicuous) [inline src/main/java/frc/robot/subsystems/intake/IntakeIO.kt]: Why are we being inconsistent in our methods names?

File: src/main/java/frc/robot/subsystems/intake/IntakeIOSim.kt (hunk 3 above)
Comment (Something-Inconspicuous) [reply src/main/java/frc/robot/subsystems/intake/IntakeIOSim.kt]: Agreed; magic numbers are bad

File: src/main/java/frc/robot/subsystems/intake/IntakeIOSim.kt [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+        inputs.intakeStatorCurrent = intakeSim.currentDrawAmps
+    }
+
+    override fun applyVoltage(voltage: Measure<Voltage>) {
+        cachedVoltage.mut_replace(voltage)
+        intakeSim.setInputVoltage(voltage into Volts)
+    }
+
+    override fun stop() {
+        applyVoltage(Volts.of(0.0))
```
Comment (Something-Inconspicuous) [inline src/main/java/frc/robot/subsystems/intake/IntakeIOSim.kt]: Is there no stop motor for the sim?

File: src/main/java/frc/robot/subsystems/intake/IntakeIO.kt [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+import edu.wpi.first.units.Voltage
+import org.littletonrobotics.junction.LogTable
+import org.littletonrobotics.junction.inputs.LoggableInputs
+
+interface IntakeIO {
+    class IntakeInputs : LoggableInputs {
+        var intakeLinearVelocity: Double = 0.0
+        var intakeSupplyVoltage: Double = 0.0
+        var intakeMotorVoltage: Double = 0.0
+        var intakeStatorCurrent: Double = 0.0
```
Comment (Something-Inconspicuous) [inline src/main/java/frc/robot/subsystems/intake/IntakeIO.kt]: USE MEASURES, PLEASE

**Review by Something-Inconspicuous — CHANGES_REQUESTED** [review]
Measures are the main thing

File: src/main/java/frc/robot/subsystems/intake/IntakeIOSim.kt (hunk 3 above)
Comment (falOn-Dev) [reply src/main/java/frc/robot/subsystems/intake/IntakeIOSim.kt]: Gearing, and MOI, ill add those to the constructors, oops

File: src/main/java/frc/robot/subsystems/intake/IntakeIONeo.kt (hunk 1 above)
Comment (falOn-Dev) [reply src/main/java/frc/robot/subsystems/intake/IntakeIONeo.kt]: Maximum safe limit for a neo 550

File: src/main/java/frc/robot/subsystems/intake/IntakeIONeo.kt (hunk 2 above)
Comment (falOn-Dev) [reply src/main/java/frc/robot/subsystems/intake/IntakeIONeo.kt]: Converting from angular to linear

File: src/main/java/frc/robot/subsystems/intake/IntakeIOSim.kt (hunk 4 above)
Comment (falOn-Dev) [reply src/main/java/frc/robot/subsystems/intake/IntakeIOSim.kt]: Converting from angular to linear

File: src/main/java/frc/robot/subsystems/intake/IntakeIOSim.kt (hunk 5 above)
Comment (falOn-Dev) [reply src/main/java/frc/robot/subsystems/intake/IntakeIOSim.kt]: Converting from angular to linear

File: src/main/java/frc/robot/subsystems/intake/IntakeIO.kt [hunk 10]
```diff
@@ -0,0 +1,49 @@
...
+import edu.wpi.first.units.Voltage
+import org.littletonrobotics.junction.LogTable
+import org.littletonrobotics.junction.inputs.LoggableInputs
+
+interface IntakeIO {
+    class IntakeInputs : LoggableInputs {
+        var intakeLinearVelocity: MutableMeasure<Velocity<Distance>> = MutableMeasure.zero(Units.MetersPerSecond)
+        var intakeSupplyVoltage: MutableMeasure<Voltage> = MutableMeasure.zero(Units.Volts)
+        var intakeMotorVoltage: MutableMeasure<Voltage> = MutableMeasure.zero(Units.Volts)
+        var intakeStatorCurrent: MutableMeasure<Current> = MutableMeasure.zero(Units.Amps)
```
Comment (Something-Inconspicuous) [inline src/main/java/frc/robot/subsystems/intake/IntakeIO.kt]: Sorry for not mentioning this the first time, but the "intake" prefixing all of these variables is redundant. Please remove to match conventions of other io layers.

**Review by Something-Inconspicuous — CHANGES_REQUESTED** [review]
Small thing I missed the first time, sorry.

**Review by Something-Inconspicuous — APPROVED** [review]

**Pull request merged**